addons/l10n_in/models/res_partner.py

```diff
@@ -2,16 +2,55 @@
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 from odoo import api, fields, models, _
-from odoo.exceptions import ValidationError
 
 
 class ResPartner(models.Model):
     _inherit = 'res.partner'
 
-    # Use in view attrs. Need to required state_id if Country is India.
-    l10n_in_country_code = fields.Char(related="country_id.code", string="Country code")
+    l10n_in_gst_treatment = fields.Selection([
+        ('regular', 'Registered Business - Regular'),
+        ('composition', 'Registered Business - Composition'),
+        ('unregistered', 'Unregistered Business'),
+        ('consumer', 'Consumer'),
+        ('overseas', 'Overseas'),
+        ('special_economic_zone', 'Special Economic Zone'),
+        ('deemed_export', 'Deemed Export'),
+    ],string="GST Treatment")
+    l10n_in_pan_number = fields.Char(string="PAN Number")
 
-    @api.constrains('vat', 'country_id')
-    def l10n_in_check_vat(self):
-        for partner in self.filtered(lambda p: p.commercial_partner_id.country_id.code == 'IN' and p.vat and len(p.vat) != 15):
-            raise ValidationError(_('The GSTIN [%s] for partner [%s] should be 15 characters only.') % (partner.vat, partner.name))
+    @api.onchange('company_type')
+    def onchange_company_type(self):
+        res = super().onchange_company_type()
+        if self.country_id and self.country_id.code == 'IN':
+            self.l10n_in_gst_treatment = (self.company_type == 'company') and 'regular' or 'consumer'
+        return res
+
+    @api.onchange('country_id')
+    def _onchange_country_id(self):
+        res = super()._onchange_country_id()
+        if self.country_id and self.country_id.code != 'IN':
+            self.l10n_in_gst_treatment = 'overseas'
+        elif self.country_id and self.country_id.code == 'IN':
+            self.l10n_in_gst_treatment = (self.company_type == 'company') and 'regular' or 'consumer'
+        return res
+
+    @api.onchange('vat')
+    def onchange_vat(self):
+        if self.vat and self.check_vat_in(self.vat):
+            self.l10n_in_pan_number = self.vat[2:13]
+            state_id = self.env['res.country.state'].search([('l10n_in_tin', '=', self.vat[:2])], limit=1)
+            if state_id:
+                self.state_id = state_id
+        else:
+            self.l10n_in_pan_number = ''
+
+    @api.onchange('l10n_in_gst_treatment')
+    def onchange_l10n_in_gst_treatment(self):
+        if self.l10n_in_gst_treatment:
+            self.vat = False
+            self.l10n_in_pan_number = False
```

Comment:
This is here because when user change treatment then the same GSTIN is not applied but you are right if the customer change treatment then this annoying the user.
So, I am remove this.